### DISCUSSION TITLE
Batched updates

### DISCUSSION THREAD

**Discussion by huczk** [pmndrs/zustand#1803] [Ideas]
Zustand is great, but it does not currently support one feature that would help me a lot - batch updates. Batch updates are useful when multiple state changes need to be made within a single action or event.

One approach to adding batch update support is to add a new method to the store object, such as `batchUpdate()`, that accepts a callback function. This function would contain all the state updates that need to be made, and Zustand would ensure that they are all executed together as a single atomic transaction or pause listeners and trigger them after `batchUpdate` will finish. Another approach would be to add a middleware.

**Comment by dai-shi** [comment]
We made a decision to rely on React for batching. It should cover most cases.
If you need batching, please feel free to create a third-party library to provide a util function or a middleware.

**Reply by huczk** [reply]
The most important in batching for me, is to stop listeners from running, I looked into the code and to do that, I need to access listeners that are called while updating state in `src/vanilla.ts:79`, and can't do it from utility function or a middleware.

**Reply by dai-shi** [reply]
Middleware can do anything. (Even a util can mutate store API.)

I would do a simpler solution.

```js
const useFooStore = create((set) => ...)
const batchUpdateFooStore = (callback) => ...
// usage
batchUpdateFooStore((set) => ...)
```

Anyway, there can be many solutions. Hope you find something you like.
Feel free to let me know if you need more help.

**Reply by huczk** [reply]
Ok, I will try to do it, if any question will come up, I will reach You out in this discussion :)

**Reply by denganjia** [reply]
> Ok, I will try to do it, if any question will come up, I will reach You out in this discussion :)

hi, more than a year has passed, is there any progress

**Reply by huczk** [reply]
I ended up with forking the package, and added function `batch(callback: () => void)` which block listeners calling until callback finish

**Reply by babacarcissedia** [reply]
This is exactly what I am looking for while working on my gps tracking solution. Need to send many state updates as one as a result of an event.
